File: vuln-sandbox/python/cwe-327/jwt-weak-algorithm-safe.py
# SAFE: jwt-weak-algorithm — RS256 asymmetric algorithm used instead of weak HS256
# Rule: CryptoJwtWeakAlgorithm | CWE-327 | Expected: TrueNegative

import os
import jwt
from flask import Flask, request, jsonify

app = Flask(__name__)

# SAFE: RS256 asymmetric algorithm; private key signs, public key verifies
PRIVATE_KEY = open(os.environ['JWT_PRIVATE_KEY_PATH']).read()
PUBLIC_KEY = open(os.environ['JWT_PUBLIC_KEY_PATH']).read()


@app.route('/token', methods=['POST'])
def issue_token():
    user_id = request.json.get('user_id')
    token = jwt.encode({'sub': user_id}, PRIVATE_KEY, algorithm='RS256')
    return jsonify({'token': token})


@app.route('/verify')
def verify():
    token = request.headers.get('Authorization', '').removeprefix('Bearer ')
    try:
        payload = jwt.decode(token, PUBLIC_KEY, algorithms=['RS256'])
    except jwt.InvalidTokenError as e:
        return jsonify({'error': str(e)}), 401
    return jsonify({'user': payload})


if __name__ == '__main__':
    app.run()

File: vuln-sandbox/node/cwe-327/jwt-weak-algorithm-safe.js
// SAFE: jwt-weak-algorithm — strong RS256 algorithm used instead of weak HS256 with shared secret
// Rule: CryptoJwtWeakAlgorithm | CWE-327 | Expected: TrueNegative

const express = require('express');
const jwt = require('jsonwebtoken');
const fs = require('fs');

const app = express();

// SAFE: RS256 asymmetric algorithm used; private key signs, public key verifies
const privateKey = fs.readFileSync(process.env.JWT_PRIVATE_KEY_PATH);
const publicKey = fs.readFileSync(process.env.JWT_PUBLIC_KEY_PATH);

app.post('/token', (req, res) => {
  const token = jwt.sign({ sub: req.body.userId }, privateKey, {
    algorithm: 'RS256',
    expiresIn: '1h',
  });
  res.json({ token });
});

app.get('/verify', (req, res) => {
  const token = req.headers.authorization?.split(' ')[1];
  jwt.verify(token, publicKey, { algorithms: ['RS256'] }, (err, decoded) => {
    if (err) return res.status(401).json({ error: 'Invalid token' });
    res.json({ user: decoded });
  });
});

app.listen(3000);
